nixos/modules/services/system/self-deploy.nix

```diff
@@ -0,0 +1,167 @@
+    in
+      lib.concatStrings (lib.mapAttrsToList toArg args);
+
+  isPathType = x: lib.strings.isCoercibleToString x && builtins.substring 0 1 (toString x) == "/";
+
+in {
+  options.services.self-deploy = {
+    enable = lib.mkEnableOption "self-deploy";
+
+    nixFile = lib.mkOption {
+      type = lib.types.path;
+
+      default = "/default.nix";
+
+      description = ''
+      Path to nix file in repository. Leading '/' refers to root of
+      git repository.
+      '';
+    };
+
+    nixAttribute = lib.mkOption {
+      type = lib.types.str;
+
+      description = ''
+      Attribute of `nixFile` that builds the current system.
+      '';
+    };
+
+    nixArgs = lib.mkOption {
+      type = lib.types.attrs;
+
+      default = {};
+
+      description = ''
+      Arguments to `nix-build` passed as `--argstr` or `--arg` depending on
+      the type.
+      '';
+    };
+
+    switchCommand = lib.mkOption {
+      type = lib.types.enum [ "boot" "switch" "dry-activate" "test" ];
+
+      default = "switch";
+
+      description = ''
+      The `switch-to-configuration` subcommand used.
+      '';
+    };
+
+    repository = lib.mkOption {
+      type = with lib.types; oneOf [ path str ];
+
+      description = ''
+      The repository to fetch from. Must be properly formatted for git.
+
+      If this value is set to a path (must begin with `/`) then it's
+      assumed that the repository is local and the resulting service
+      won't wait for the network to be up.
+
+      If the repository will be fetched over SSH, you must add an
+      entry to `programs.ssh.knownHosts` for the SSH host for the fetch
+      to be successful.
+      '';
+    };
+
+    sshKeyFile = lib.mkOption {
+      type = with lib.types; nullOr path;
+
+      default = null;
+
+      description = ''
+      Path to SSH private key used to fetch private repositories over
+      SSH.
+      '';
+    };
+
+    branch = lib.mkOption {
+      type = lib.types.str;
+
+      default = "master";
+
+      description = ''
+      Branch to track
+
+      Technically speaking any ref can be specified here, as this is
+      passed directly to a `git fetch`, but for the use-case of
+      continuous deployment you're likely to want to specify a branch.
+      '';
+    };
+
+    startAt = lib.mkOption {
+      type = with lib.types; either str (listOf str);
+
+      default = "hourly";
+
+      description = ''
+      The schedule on which to run the `self-deploy` service. Format
+      specified by `systemd.time 7`.
+
+      This value can also be a list of `systemd.time 7` formatted
+      strings, in which case the service will be started on multiple
+      schedules.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.self-deploy = {
+      wantedBy = [ "multi-user.target" ];
+
+      requires = lib.mkIf (!(isPathType cfg.repository)) [ "network-online.target" ];
```

Comment:
If the script runs by a cron we don't really need network at startup, right?

Reply:
It's probably best to make it explicit that network is needed regardless. Probably will never be necessary, but there's something to be said for correctness.

Reply:
Also I am not sure if this requires network target additionally.